src/team_lib/mhba/xccl_mhba_socket_comm.c

```diff
@@ -0,0 +1,277 @@
+
+    *shared_cmd_fd = *p;
+    *shared_pd_handle = handles[0];
+
+    return XCCL_OK;
+}
+
+static xccl_status_t client_recv_data(int* shared_cmd_fd,uint32_t *shared_pd_handle)
+{
+    int sock;
+    struct sockaddr_storage sockaddr = {};
+    struct sockaddr_un *addr;
+
+    sock = socket(PF_LOCAL, SOCK_STREAM, 0);
+    if (sock == -1)
+    {
+        xccl_mhba_error("Failed to create client socket errno %d", errno);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    addr = (struct sockaddr_un *)&sockaddr;
+    addr->sun_family = AF_UNIX;
+    strncpy(addr->sun_path, SOCKPATH, sizeof(addr->sun_path));
+    addr->sun_path[sizeof(addr->sun_path) -1 ] = '\0';
+
+    if (connect(sock, (struct sockaddr *)addr, SUN_LEN(addr)) == -1) {
+        xccl_mhba_error("Failed to connect client socket errno %d", errno);
+        goto fail;
+    }
+    if (do_recvmsg(sock, shared_cmd_fd, shared_pd_handle) != XCCL_OK) {
+        xccl_mhba_error("Failed to recv msg");
+        goto fail;
+    }
+
+    if (close(sock) == -1){
+        xccl_mhba_error("Failed to close client socket errno %d", errno);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+    return XCCL_OK;
+
+fail:
+    if (close(sock) == -1){
+        xccl_mhba_error("Failed to close client socket errno %d", errno);
+    }
+    return XCCL_ERR_NO_MESSAGE;
+}
+
+static xccl_status_t server_send_data(int command_fd, uint32_t pd_handle, int num_of_connections)
+{
+    int sock, i, num_of_curr_connections = 0;
+
+    struct sockaddr_storage storage= {};
+    struct sockaddr_un *addr;
+
+    connection_t connection[num_of_connections];
+    pthread_t thread[num_of_connections];
+
+    sock = socket(PF_LOCAL, SOCK_STREAM, 0);
+    if (sock == -1)
+    {
+        xccl_mhba_error("Failed to create server socket errno %d", errno);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    addr = (struct sockaddr_un *)&storage;
+    addr->sun_family = AF_UNIX;
+    strncpy(addr->sun_path, SOCKPATH, sizeof(addr->sun_path));
+    addr->sun_path[sizeof(addr->sun_path) -1 ] = '\0';
+
+    if (bind(sock, (struct sockaddr *)addr, sizeof(struct sockaddr_un)) == -1) {
+        xccl_mhba_error("Failed to bind server socket errno %d", errno);
+        goto fail;
+    }
+    if (listen(sock, num_of_connections) == -1) {
+        xccl_mhba_error("Failed to listen to server socket errno %d", errno);
+        goto listen_fail;
+    }
+
+    while (num_of_curr_connections < num_of_connections)
+    {
+        /* accept incoming connections */
+        connection[num_of_curr_connections].sock = accept(sock, NULL, 0);
+        connection[num_of_curr_connections].fd = command_fd;
+        connection[num_of_curr_connections].pd_handle = pd_handle;
+        if (connection[num_of_curr_connections].sock != -1)
+        {
+            /* start a new thread but do not wait for it */
+            pthread_create(&thread[num_of_curr_connections], 0, do_sendmsg, (void *)
```

Comment:
i don't see any sense in spawning a thread and immediately after waiting for it to complete with pthread_join. what is it ?

Reply:
correct, my mistake. fixed.